2.0.0-Data-Model.md

```diff
@@ -12,22 +12,79 @@ We are starting with a read-only API due to potentially different views and appr
-
-For example, if a WES server was asked to process `trs://trs.example.org/123456323`, it would know that it could issue a GET request to `https://trs.example.org/api/ga4gh/trs/v2/tools/123456323` to learn how to fetch that object.
+#### TRS Tool and TRS Tool Version IDs
+
+Each implementation of TRS can choose its own identifier scheme, as long as it
+follows these guidelines:
+
+* TRS Tool and TRS Tool Version IDs are strings made up of uppercase and
+  lowercase letters, decimal digits, hyphen, period, underscore and tilde
+  (`[A-Za-z0-9.-_~]`). See [RFC 3986 §
+  2.3](https://datatracker.ietf.org/doc/html/rfc3986#section-2.3). TRS Tool and
+  TRS Tool Version IDs MAY further contain percent characters (`%`) whenever
+  they are used in API calls, but only if they were introduced through
+  percent-encoding of any non-valid characters (see next bullet point).
+* TRS Tool and TRS Tool Version IDs can contain other characters, but they MUST
+  be percent-encoded (see [RFC 3986 §
+  2.4](https://datatracker.ietf.org/doc/html/rfc3986#section-2.4)) into valid
+  TRS Tool and TRS Tool Version IDs as per the previous rule whenever they are
+  used in API calls. This is because non-encoded IDs may interfere with the
+  interpretation of routes, e.g., for the `/tools/{id}/versions/{version_id}`
+  endpoint.
+* Any given TRS Tool or TRS Tool Version ID MUST always identify the same
+  resource (tool or tool version, respectively) on a given TRS implementation.
+  This constraint aids with reproducibility.
+* TRS implementations MAY have more than one TRS Tool or TRS Tool Version ID
+  mapping to the same resource (tool or tool version, respectively).
 
-This recommendation is intended to mirror the discussion that went into the [DRS URI scheme](https://ga4gh.github.io/data-repository-service-schemas/preview/develop/docs/#_drs_uris).
+#### TRS URIs
 
-For informational purposes, we recommend that TRS implementations add themselves to https://github.com/ga4gh/tool-registry-service-schemas/blob/develop/registry.json to provide for the possibility of creating a global indexing service and to allow others to more easily discover a TRS implementation. 
+To conveniently pass content references to TRS resources, e.g., to advise a
+[WES](https://github.com/ga4gh/workflow-execution-service-schemas) or
+[TES](https://github.com/ga4gh/task-execution-schemas) service which tool
+(version) to use, we define a URI syntax for TRS-accessible content. Strings of
+the form `trs://<server>/<id>` (unversioned) and
+`trs://<server>/<id>/<version_id>` (versioned) mean _“you can fetch the content
+with TRS Tool ID `<id>` and, if provided, TRS Tool Version ID `<version_id>`
+from the TRS server at `<server>` "_.
+
+For example, if a TRS client was asked to process
+`trs://trs.example.org/tool_ABC/v1.2.3`, it would know that it could, e.g.,
+issue `GET` requests to
+`https://trs.example.org/api/ga4gh/trs/v2/tools/tool_ABC` and
```

Comment:
I think versioning APIs is an important topic and I think there's most definitely a need for discussion and harmonization across GA4GH APIs (e.g., Stian mentioned HATEOAS in #160), but as this PR does not introduces the issue nor is designed to address it, I don't think it should be blocked by it.

Besides, I don't think that TRS URIs, as identifiers of resources, should be concerned with API versioning at all: IMO, when bumping API versions, care must be taken (both on the spec- and implementation-side) not to introduce ambiguities in the way that the different responses in, e.g., TRS `v2` and `v3` can be interpreted by a client. And if changes to the internal representation of resources are required that would break backward compatibility, then I think those resources should be cloned and given new identifiers, such that reproducibility is not negatively affected. As long as these considerations/constraints are honored, it shouldn't matter which API version is used to access a given resource.

Reply:
In any case, I have clarified this/addressed these concerns now in lines 56/57 and 69-72

Reply:
My question was not about API/schema versioning, which is a very complex subject and I'm no expert in. It was about the example that if you have a URI `trs://trs.example.org/tool_ABC/v1.2.3`, it says you can fetch the content `https://trs.example.org/api/ga4gh/trs/v2/tools/tool_ABC` (note the `v2` in the path).

If TRS v3 comes along, and a new TRS implementation only supports that v3,  then the `v2` in the path means they would return a 404. Maybe that's OK, or maybe the requirement is that all TRS implementations need to implement v2 and above. Otherwise a client will need to iterate over paths with all potential versions.

Maybe I'm overthinking this, and it is a real edge case.

Reply:
I do think this _is_ an API/schema versioning topic because the API version is precisely what `v2` refers to. In any case, there anyway needs to be some sort of negotiation between a service and a client, because not every client will implement every version of the API specs, and so there will be cases where clients and servers won't match, regardless of whether the client will iterate or whether we specify the version in the TRS URI. One possibility to facilitate this negotation would be to define the supported API versions in the service info, but of course, the Service Info API is versioned, too...

I agree this is an important issue and needs to be discussed on a larger scale within the Cloud WS in order to come up with a sensible strategy that covers DRS and TRS URIs as well as other artefacts/problems with API versioning. A related issue I faced is also that the path (e.g., `ga4gh/trs/v2`) defined in the schema may not always be possible or desirable to implement. For example, as far as I know, Dockstore does not. The path info could perhaps also be defined in the service info for any given service (although the same problem applies: how do you get to the service info in the first place; perhaps via the service registry...). 

Still, I don't think this PR makes this problem worse than it is already, given the current definitions of DRS and TRS URIs, so I don't think it should block.

Reply:
> I do think this is an API/schema versioning topic

Fair. I was thinking more of different schemas in different versions of the API when I made my comment, but you're right.

> Still, I don't think this PR makes this problem worse than it is already, given the current definitions of DRS and TRS URIs

I think the explicitness that you can take a URI and translate it to a specific path is a new wrinkle, at least for TRS. You're right about DRS -- I hadn't looked until now and they do the same thing, with `v1` in the path.

> so I don't think it should block.

I had already approved, so it is not blocked.

Reply:
So @denis-yuen, others: shall we bring up this important discussion to the Cloud WS?

Also perhaps to see if people agree that it might be useful to specify TRS URIs as inputs to WES (to access workflows) and inside workflows (to access containers through WES/TES implementations)? Because this might require changes on WES and TES specs as well, without which TRS URIs might not be all that useful.

Reply:
Proposal at Cloud WS was to break this discussion into a separate ticket https://github.com/ga4gh/tool-registry-service-schemas/issues/221